### PR TITLE
[aot] Build cross compilers targeting x64/x86 hosted on osx-arm64

### PR DESCRIPTION
**Pull request by lambdageek**
This PR allows an `osx-arm64` or `osx-x64` machine to produce a `mono-aot-cross` compiler that will run on the following host and create AOT images for the following targets:

| Host | Target | Test App runs |
|--------|--------|--------|
| osx-arm64 | iossimulator-x64 | ✅  |
| osx-arm64 | android-x64 | ✅   |
| osx-arm64 | android-x86 | ✅  - depends on a hack for https://github.com/dotnet/runtime/issues/89039 | 

Additionally a `linux-x64` or `linux-arm64` machine can build a `mono-aot-cross` compiler for these host/target pairs:

| Host | Target | Test App runs |
|--------|--------|--------|
| linux-arm64 | android-x64 | :question: |
| linux-arm64 | linux_musl-x64 | :question: |
| linux-arm64 | android-x86 | :question: |

Related to https://github.com/dotnet/runtime/issues/82495

### CODE REVIEW

File: src/mono/mono/mini/exceptions-amd64.c [hunk 1]
```diff
@@ -904,7 +904,7 @@ altstack_handle_and_restore (MonoContext *ctx, MonoObject *obj, guint32 flags)
 void
 mono_arch_handle_altstack_exception (void *sigctx, MONO_SIG_HANDLER_INFO_TYPE *siginfo, gpointer fault_addr, gboolean stack_ovf)
 {
-#if defined(MONO_ARCH_USE_SIGACTION)
```
Comment (vargaz) [inline src/mono/mono/mini/exceptions-amd64.c:907]: We usually use !defined(MONO_CROSS_COMPILE) to avoid compiling these.

**Comment by lambdageek** [comment]
host: osx-arm64 target: android-x64 also works with this PR

I probably will need to try host: linux-arm64 target: android-x64 separately since some of the changes here were in the `__APPLE__` ifdef...

**Comment by lambdageek** [comment]
host osx-arm64 target android-x86 is also now part of this PR

**Comment by lambdageek** [comment]
Getting the linux-arm64 stuff working depends on me re-learning how to build stuff using Docker images

**Review by vargaz — APPROVED** [review]

**Comment by lambdageek** [comment]
Oh, the android-x86 build depends on a hack to work around https://github.com/dotnet/runtime/issues/89039

```diff
diff --git a/eng/native/tryrun.cmake b/eng/native/tryrun.cmake
index 97947fe8879..a31a1de3f23 100644
--- a/eng/native/tryrun.cmake
+++ b/eng/native/tryrun.cmake
@@ -48,7 +48,7 @@ elseif(EXISTS ${CROSS_ROOTFS}/boot/system/develop/headers/config/HaikuConfig.h)
 endif()
 
 if(DARWIN)
-  if(TARGET_ARCH_NAME MATCHES "^(arm64|x64)$")
+  if(TARGET_ARCH_NAME MATCHES "^(arm64|x64|x86)$")
     set_cache_value(FILE_OPS_CHECK_FERROR_OF_PREVIOUS_CALL_EXITCODE 1)
     set_cache_value(HAS_POSIX_SEMAPHORES_EXITCODE 1)
     set_cache_value(HAVE_BROKEN_FIFO_KEVENT_EXITCODE 1)
```

**Comment by lambdageek** [comment]
Attempting host linux-arm64 target android-x64 using

```
docker run -e ROOTFS_DIR=/crossrootfs/arm64 -w `pwd` --platform linux/amd64 -v `pwd`:`pwd` -it mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-cross-arm64-20220427171722-6e40d49 ./build.sh -subset mono+packs -c release -arch arm64 -ci  /p:MonoCrossAOTTargetOS=Android /p:SkipMonoCrossJitConfigure=true /p:BuildMonoAOTCrossCompilerOnly=true /bl
```

based on the instructions in https://github.com/dotnet/runtime/issues/74175#issuecomment-1228860772

**Comment by directhex** [comment]
Android Linux host repro command is:

```
docker run -e ROOTFS_DIR=/crossrootfs/arm64 -w $(pwd) --platform linux/amd64 -v $(pwd):$(pwd) -it mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-cross-arm64 ./build.sh -subset mono+packs -c release -arch arm64 -ci  /p:MonoCrossAOTTargetOS=android /p:SkipMonoCrossJitConfigure=true /p:BuildMonoAOTCrossCompilerOnly=true /bl
```

Relies on offset files from the most recent build of https://dev.azure.com/dnceng-public/public/_build?definitionId=129&branchFilter=331 being in `artifacts/obj/mono/offsetfiles/` (e.g. `artifacts/obj/mono/offsetfiles/android.x64.Release/cross/offsets-x64-android.h` should exist)

**Comment by directhex** [comment]
Linux fix:

```
diff --git a/src/mono/mono/utils/mono-context.h b/src/mono/mono/utils/mono-context.h
index aa444a2f980..9ce932575b1 100644
--- a/src/mono/mono/utils/mono-context.h
+++ b/src/mono/mono/utils/mono-context.h
@@ -47,7 +47,12 @@ typedef __uint128_t MonoContextSimdReg;
 #endif
 #elif defined(__linux__) && defined(__GLIBC__)
 #define MONO_HAVE_SIMD_REG
+#if !defined(MONO_CROSS_COMPILE)
 typedef struct _libc_xmmreg MonoContextSimdReg;
+#else
+/* if building for an arm64 host machine, a cross compiler that produces x64 modules, fake a MonoContextSimdReg */
+typedef __uint128_t MonoContextSimdReg;
+#endif
 #elif defined(HOST_WIN32)
 #define MONO_HAVE_SIMD_REG
 //#define MONO_HAVE_SIMD_REG_AVX
```

**Comment by directhex** [comment]
I think this covers yaml changes https://gist.github.com/directhex/6af1d74199084a992dad250294aa876e

Workloads generally need some finesse to deal with all the new native arm64 stuff we have going in lots of places

**Comment by lambdageek** [comment]
@directhex Applied both of your patches, thanks! The x64 CI machines should be building the arm64 packages now, I believe.

@steveisok @akoeplinger please take a look

**Review by akoeplinger — APPROVED** [review]

**Pull request merged**